crates/red_knot_python_semantic/src/semantic_index/use_def/symbol_state.rs

```diff
@@ -0,0 +1,362 @@
+//! Track visible definitions of a symbol, and applicable constraints per definition.
+//!
+//! These data structures operate entirely on scope-local newtype-indices for definitions and
+//! constraints, referring to their location in the `all_definitions` and `all_constraints`
+//! indexvecs in [`super::UseDefMapBuilder`].
+//!
+//! We need to track arbitrary associations between definitions and constraints, not just a single
+//! set of currently dominating constraints (where "dominating" means "control flow must have
+//! passed through it to reach this point"), because we can have dominating constraints that apply
+//! to some definitions but not others, as in this code:
+//!
+//! ```python
+//! x = 1 if flag else None
+//! if x is not None:
+//!     if flag2:
+//!         x = 2 if flag else None
+//!     x
+//! ```
+//!
+//! From the point of view of the final use of `x`, the `x is not None` constraint is "currently
+//! visible" (in the sense that control flow must have passed through it to reach this point), but
+//! it applies only to the first definition of `x`, not the second, so `None` is a possible value
+//! for `x`.
+//!
+//! And we can't just track an index into a list of dominating constraints for each definition,
+//! either, because we can have definitions which are still visible, but subject to constraints
+//! that are no longer dominating, as in this code:
+//!
+//! ```python
+//! x = 0
+//! if flag1:
+//!     x = 1 if flag2 else None
+//!     assert x is not None
+//! x
+//! ```
+//!
+//! From the point of view of the final use of `x`, the `x is not None` constraint no longer
+//! dominates, but it does dominate the `x = 1 if flag2 else None` definition, so we have to keep
+//! track of that.
+//!
+//! The data structures used here ([`BitSet`] and [`BitSetArray`]) optimize for keeping all data
+//! inline (avoiding lots of scattered allocations) in small-to-medium cases, and falling back to
+//! heap allocation to be able to scale to arbitrary numbers of definitions and constraints when
+//! needed.
+use super::bitset::{BitSet, BitSetArray, BitSetArrayIterator, BitSetIterator};
+use ruff_index::newtype_index;
+
+/// A newtype-index for a definition in a particular scope.
+#[newtype_index]
+pub(super) struct ScopedDefinitionId;
+
+/// A newtype-index for a constraint expression in a particular scope.
+#[newtype_index]
+pub(super) struct ScopedConstraintId;
+
+/// Can reference this * 64 total definitions inline; more will fall back to the heap.
+const INLINE_DEFINITION_BLOCKS: usize = 1;
+
+/// A [`BitSet`] of [`ScopedDefinitionId`], representing visible definitions of a symbol in a scope.
+type Definitions = BitSet<INLINE_DEFINITION_BLOCKS>;
+type DefinitionsIterator<'a> = BitSetIterator<'a, INLINE_DEFINITION_BLOCKS>;
+
+/// Can reference this * 64 total constraints inline; more will fall back to the heap.
+const INLINE_CONSTRAINT_BLOCKS: usize = 1;
```

Comment:
Same here, we probably want to use a number higher than 1

Reply:
I initially had a higher number and switched to 1 because it performed best. In this case I'm not sure the argument for using a bigger number is as strong, because I suspect an array of 1 `BitSet` is already as large as a `Vec`? But I'll double check both the sizes and the actual performance.

Reply:
`[BitSet<1>; 1]` is 32 bytes, `Vec<BitSet>` is 24 bytes, so there's no wasted memory in leaving this at 1. Empirically 1 and 2 perform the same on the benchmark, 3 is a small regression, 4 is a large regression, so I'll set this to 2 for now.